### ISSUE TITLE
[Question] Some questions about namespace bundle splitting

### ISSUE THREAD

**Issue by JevonQ** [apache/pulsar#3419]
According to the definition of bundle and relationship between a topic and a bundle, once a topic is created the hash of its name will be constant. If there is a splitting happens on a bundle, the hash of some of the topics that belong to the bundle will fall into the new bundle. If the new bundle is assigned to another broker, then the load for those topics will be shed to another broker. But if there is always heavy load on one certain topic, splitting on the bundle does not help to spread the load on that topic, is it correct?

If that's the case, how to solve it?

**Comment by merlimat** [comment]
@JevonQ That's correct. Topics are the lowest unit of assignments to brokers and a topic is served by only 1 broker at a given point in time.

Keep in mind that Pulsar supports partitioned topics too (http://pulsar.apache.org/docs/en/admin-api-partitioned-topics.html#docsNav). If a topic is partitioned, it will be basically composed of multiple partitions (which are seen by the broker as independent topics). Different partition can thus be split into different namespace bundles.

To summarize: if the load on a topic is expected to be "high" (more than what 1 single node can comfortably sustain), then you should enable partitioning.

**Comment by JevonQ** [comment]
@merlimat I meant to enable partioning on the  topic, if the partitioned-topic is overloaded, how to solve it?

**Comment by lovelle** [comment]
Hi! Afaik the default strategy for partitioned topics is round-robin, therefore you shouldn't have just 1 partition with much more load than the rest of the partitions, of course more partitions on a topic would result in a more finer grained load handling if you have enough broker to handle those partitions.

**Comment by merlimat** [comment]
@JevonQ in Pulsar there are "topics" and "partitioned-topics".

Partitioned topics are basically just a composition of non-partitioned topics. So, if you have a namespace with, say, 2 bundles and  1 partitioned topic composed of 16 partitions, you should expect that each bundle will contain roughly the same number of partitions (depending on where the hashing falls).

Basically, the bundles can be kept splitting in half (if the load is high) until we reach the point where there's just 1 single partition left.

**Comment by JevonQ** [comment]
> @JevonQ in Pulsar there are "topics" and "partitioned-topics".
> 
> Partitioned topics are basically just a composition of non-partitioned topics. So, if you have a namespace with, say, 2 bundles and 1 partitioned topic composed of 16 partitions, you should expect that each bundle will contain roughly the same number of partitions (depending on where the hashing falls).
> 
Yes, I understand this.
> Basically, the bundles can be kept splitting in half (if the load is high) until we reach the point where there's just 1 single partition left.
Yes, my question is about this. If the bundle only contains one partition and the partition is still overloaded, is there a way to solve it. Increasing partition does not work for me cause I use key hash as the routing mode while producing the messages.

**Comment by sijie** [comment]
> Yes, my question is about this. If the bundle only contains one partition and the partition is still overloaded, is there a way to solve it? 

Currently partition is the finest granularity in Pulsar. We can't split it now. 

> Increasing partition does not work for me cause I use key hash as the routing mode while producing the messages.

Does the new subscription mode `key_shared` address your problem?
